File: k3_primary_evidence.py
#!/usr/bin/env python3
"""Single audited access point for Kimi K3's disclosed architecture counts.

The launch page rounded K3 to ``2.8T`` and ``104B``.  The subsequently released
technical report gives the exact Table 1 values used by analytical code:
2.78 trillion total parameters and 104.2 billion activated parameters.  Keeping
the display-rounded value separate prevents it from leaking into calibration or
backtest arithmetic.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any


ROOT = Path(__file__).resolve().parent
K3_EVIDENCE_PATH = ROOT / "sources/kimi_k3_release_evidence_2026-07-31.json"
K3_PARAMETER_SOURCE = "Kimi K3 official technical report Table 1"


def load_k3_primary_evidence() -> dict[str, Any]:
    record = json.loads(K3_EVIDENCE_PATH.read_text(encoding="utf-8"))
    k3 = record.get("kimi_k3", {})
    if (
        float(k3.get("total_parameters_b_exact", 0)) != 2780.0
        or float(k3.get("activated_parameters_b_exact", 0)) != 104.2
        or k3.get("parameter_count_disclosed") is not True
        or k3.get("activated_parameter_count_disclosed") is not True
    ):
        raise RuntimeError(
            "Kimi K3 primary evidence is missing the exact disclosed "
            "2.78T total / 104.2B activated counts"
        )
    return record


K3_PRIMARY_EVIDENCE = load_k3_primary_evidence()
K3_ARCHITECTURE = K3_PRIMARY_EVIDENCE["kimi_k3"]
K3_TOTAL_B = float(K3_ARCHITECTURE["total_parameters_b_exact"])
K3_TOTAL_T = K3_TOTAL_B / 1000.0
K3_ACTIVE_B = float(K3_ARCHITECTURE["activated_parameters_b_exact"])
K3_TOTAL_T_DISPLAY = float(K3_ARCHITECTURE["total_parameters_t_display"])

File: k3_primary_evidence.mjs
import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

const root = path.dirname(fileURLToPath(import.meta.url));
export const k3EvidencePath = path.join(
  root,
  "sources/kimi_k3_release_evidence_2026-07-31.json",
);
export const k3ParameterSource = "Kimi K3 official technical report Table 1";
export const k3Evidence = JSON.parse(await fs.readFile(k3EvidencePath, "utf8"));
export const k3Architecture = k3Evidence.kimi_k3;
export const k3TotalB = Number(k3Architecture.total_parameters_b_exact);
export const k3TotalT = k3TotalB / 1000;
export const k3ActiveB = Number(k3Architecture.activated_parameters_b_exact);
export const k3TotalTDisplay = Number(k3Architecture.total_parameters_t_display);

if (
  k3TotalB !== 2780
  || k3ActiveB !== 104.2
  || k3Architecture.parameter_count_disclosed !== true
  || k3Architecture.activated_parameter_count_disclosed !== true
) {
  throw new Error(
    "Kimi K3 primary evidence is missing the exact disclosed 2.78T total / 104.2B activated counts",
  );
}
